twine/auth.py

```diff
@@ -102,12 +173,19 @@ def username_from_keyring_or_prompt(self) -> str:
 
         return self.prompt("username", input)
 
-    def password_from_keyring_or_prompt(self) -> str:
+    def password_from_keyring_or_trusted_publishing_or_prompt(self) -> str:
         password = self.get_password_from_keyring()
         if password:
             logger.info("password set from keyring")
             return password
 
+        if self.is_pypi() and self.username == "__token__":
```

Comment:
Do we _need_ to special-case PyPI here? I don't see any reason why this shouldn't work for a non-PyPI index that has implemented the same flow otherwise, and it would be nice if we didn't have to special case every index that wants to support an OIDC-based token exchange in the future.

Reply:
There are zero of those today. There is little value in fielding lots of issue questions about why this doesn't work with Nexus/Artifactory/GitLab private registry/etc or have folks open issues thinking this is the reason they can't upload (because they saw it in the verbose/debug logs). We already have people misunderstanding that a bad request response from PyPI is a twine problem.

Reply:
Also in my experience private package indices take a good year to 10 to catch up with the rest of the world

Reply:
Fair enough.